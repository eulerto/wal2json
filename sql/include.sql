\set VERBOSITY terse

-- predictability
SET synchronous_commit = on;

DROP TABLE IF EXISTS t1;
DROP TABLE IF EXISTS t2;
DROP TABLE IF EXISTS t3;

CREATE TABLE t1 (id int PRIMARY KEY);
CREATE TABLE t2 (id int PRIMARY KEY);
CREATE TABLE t3 (id int PRIMARY KEY);
CREATE TABLE s1 (id int PRIMARY KEY);

SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');


-- Include commands in single tables

insert into t1 values (1);
insert into t2 values (2);
insert into t3 values (3);
update t1 set id = 10 where id = 1;
update t2 set id = 20 where id = 2;
update t3 set id = 30 where id = 3;
delete from t1 where id = 10;
delete from t2 where id = 20;
delete from t3 where id = 30;

SELECT data FROM pg_logical_slot_get_changes(
	'regression_slot', NULL, NULL, 'include-xids', '0', 'pretty-print', '1',
	'include-table', 't1', 'include-table', 't3');


-- Include commands on a pattern of tables

insert into t1 values (1);
insert into t2 values (2);
insert into s1 values (3);

SELECT data FROM pg_logical_slot_get_changes(
	'regression_slot', NULL, NULL, 'include-xids', '0', 'pretty-print', '1',
	'include-table', '~t');


insert into t1 values (4);
insert into t2 values (5);
insert into s1 values (6);

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'include-table', '~^.1$');


-- Exclude a table after inclusion

insert into t1 values (7);
insert into t2 values (8);
insert into t3 values (9);
insert into s1 values (10);

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'include-table', '~^t', 'exclude-table', 't2');


-- Exclude a single table

insert into t1 values (11);
insert into t2 values (12);
insert into t3 values (13);


SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'exclude-table', 't2');


-- Exclude a pattern

insert into t1 values (14);
insert into t2 values (15);
insert into t3 values (16);
insert into s1 values (17);


SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'exclude-table', '~.1');


-- Include after exclusion

insert into t1 values (18);
insert into t2 values (19);
insert into t3 values (20);
insert into s1 values (21);


SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'exclude-table', '~t', 'include-table', '~.2');


SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');
