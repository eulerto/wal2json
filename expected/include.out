\set VERBOSITY terse
-- predictability
SET synchronous_commit = on;
DROP TABLE IF EXISTS t1;
NOTICE:  table "t1" does not exist, skipping
DROP TABLE IF EXISTS t2;
NOTICE:  table "t2" does not exist, skipping
DROP TABLE IF EXISTS t3;
NOTICE:  table "t3" does not exist, skipping
CREATE TABLE t1 (id int PRIMARY KEY);
CREATE TABLE t2 (id int PRIMARY KEY);
CREATE TABLE t3 (id int PRIMARY KEY);
CREATE TABLE s1 (id int PRIMARY KEY);
SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');
 ?column? 
----------
 init
(1 row)

-- Include commands in single tables
insert into t1 values (1);
insert into t2 values (2);
insert into t3 values (3);
update t1 set id = 10 where id = 1;
update t2 set id = 20 where id = 2;
update t3 set id = 30 where id = 3;
delete from t1 where id = 10;
delete from t2 where id = 20;
delete from t3 where id = 30;
SELECT data FROM pg_logical_slot_get_changes(
	'regression_slot', NULL, NULL, 'include-xids', '0', 'pretty-print', '1',
	'include-table', 't1', 'include-table', 't3');
                         data                          
-------------------------------------------------------
 {                                                    +
         "change": [
                 {                                    +
                         "kind": "insert",            +
                         "schema": "public",          +
                         "table": "t1",               +
                         "columnnames": ["id"],       +
                         "columntypes": ["int4"],     +
                         "columnvalues": [1]          +
                 }
         ]                                            +
 }
 {                                                    +
         "change": [
         ]                                            +
 }
 {                                                    +
         "change": [
                 {                                    +
                         "kind": "insert",            +
                         "schema": "public",          +
                         "table": "t3",               +
                         "columnnames": ["id"],       +
                         "columntypes": ["int4"],     +
                         "columnvalues": [3]          +
                 }
         ]                                            +
 }
 {                                                    +
         "change": [
                 {                                    +
                         "kind": "update",            +
                         "schema": "public",          +
                         "table": "t1",               +
                         "columnnames": ["id"],       +
                         "columntypes": ["int4"],     +
                         "columnvalues": [10],        +
                         "oldkeys": {                 +
                                 "keynames": ["id"],  +
                                 "keytypes": ["int4"],+
                                 "keyvalues": [1]     +
                         }                            +
                 }
         ]                                            +
 }
 {                                                    +
         "change": [
         ]                                            +
 }
 {                                                    +
         "change": [
                 {                                    +
                         "kind": "update",            +
                         "schema": "public",          +
                         "table": "t3",               +
                         "columnnames": ["id"],       +
                         "columntypes": ["int4"],     +
                         "columnvalues": [30],        +
                         "oldkeys": {                 +
                                 "keynames": ["id"],  +
                                 "keytypes": ["int4"],+
                                 "keyvalues": [3]     +
                         }                            +
                 }
         ]                                            +
 }
 {                                                    +
         "change": [
                 {                                    +
                         "kind": "delete",            +
                         "schema": "public",          +
                         "table": "t1",               +
                         "oldkeys": {                 +
                                 "keynames": ["id"],  +
                                 "keytypes": ["int4"],+
                                 "keyvalues": [10]    +
                         }                            +
                 }
         ]                                            +
 }
 {                                                    +
         "change": [
         ]                                            +
 }
 {                                                    +
         "change": [
                 {                                    +
                         "kind": "delete",            +
                         "schema": "public",          +
                         "table": "t3",               +
                         "oldkeys": {                 +
                                 "keynames": ["id"],  +
                                 "keytypes": ["int4"],+
                                 "keyvalues": [30]    +
                         }                            +
                 }
         ]                                            +
 }
(24 rows)

-- Include commands on a pattern of tables
insert into t1 values (1);
insert into t2 values (2);
insert into s1 values (3);
SELECT data FROM pg_logical_slot_get_changes(
	'regression_slot', NULL, NULL, 'include-xids', '0', 'pretty-print', '1',
	'include-table', '~t');
                       data                       
--------------------------------------------------
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t1",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [1]     +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t2",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [2]     +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
         ]                                       +
 }
(8 rows)

insert into t1 values (4);
insert into t2 values (5);
insert into s1 values (6);
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'include-table', '~^.1$');
                       data                       
--------------------------------------------------
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t1",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [4]     +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "s1",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [6]     +
                 }
         ]                                       +
 }
(6 rows)

-- Exclude a table after inclusion
insert into t1 values (7);
insert into t2 values (8);
insert into t3 values (9);
insert into s1 values (10);
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'include-table', '~^t', 'exclude-table', 't2');
                       data                       
--------------------------------------------------
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t1",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [7]     +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t3",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [9]     +
                 }
         ]                                       +
 }
(6 rows)

-- Exclude a single table
insert into t1 values (11);
insert into t2 values (12);
insert into t3 values (13);
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'exclude-table', 't2');
                       data                       
--------------------------------------------------
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t1",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [11]    +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t3",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [13]    +
                 }
         ]                                       +
 }
(6 rows)

-- Exclude a pattern
insert into t1 values (14);
insert into t2 values (15);
insert into t3 values (16);
insert into s1 values (17);
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'exclude-table', '~.1');
                       data                       
--------------------------------------------------
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t2",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [15]    +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t3",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [16]    +
                 }
         ]                                       +
 }
(6 rows)

-- Include after exclusion
insert into t1 values (18);
insert into t2 values (19);
insert into t3 values (20);
insert into s1 values (21);
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL,
	'include-xids', '0', 'pretty-print', '1', 'skip-empty-xacts', '1',
	'exclude-table', '~t', 'include-table', '~.2');
                       data                       
--------------------------------------------------
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "t2",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [19]    +
                 }
         ]                                       +
 }
 {                                               +
         "change": [
                 {                               +
                         "kind": "insert",       +
                         "schema": "public",     +
                         "table": "s1",          +
                         "columnnames": ["id"],  +
                         "columntypes": ["int4"],+
                         "columnvalues": [21]    +
                 }
         ]                                       +
 }
(6 rows)

SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');
 ?column? 
----------
 stop
(1 row)

